src/PostgREST/QueryBuilder.hs

```diff
@@ -145,9 +134,54 @@ requestToQuery schema _ (DbMutate (Delete mainTbl logicForest returnings)) =
   where
     qi = QualifiedIdentifier schema mainTbl
 
-unquoted :: JSON.Value -> Text
-unquoted (JSON.String t) = t
-unquoted (JSON.Number n) =
-  toS $ formatScientific Fixed (if isInteger n then Just 0 else Nothing) n
-unquoted (JSON.Bool b) = show b
-unquoted v = toS $ JSON.encode v
+requestToCallProcQuery :: QualifiedIdentifier -> [PgArg] -> Bool -> Maybe PreferParameters -> SqlQuery
+requestToCallProcQuery qi pgArgs returnsScalar preferParams =
+  unwords [
+    "WITH",
+    argsCTE,
+    sourceBody ]
+  where
+    paramsAsSingleObject    = preferParams == Just SingleObject
+    paramsAsMulitpleObjects = preferParams == Just MultipleObjects
+
+    (argsCTE, args)
+      | null pgArgs = (ignoredBody, "")
+      | paramsAsSingleObject = ("pgrst_args AS (SELECT NULL)", "$1::json")
+      | otherwise = (
+          unwords [
+            normalizedBody <> ",",
+            "pgrst_args AS (",
+              "SELECT * FROM json_to_recordset(" <> selectBody <> ") AS _(" <> fmtArgs (\a -> " " <> pgaType a) <> ")",
+            ")"]
+         , if paramsAsMulitpleObjects
+             then fmtArgs (\a -> " := pgrst_args." <> pgFmtIdent (pgaName a))
+             else fmtArgs (\a -> " := (SELECT " <> pgFmtIdent (pgaName a) <> " FROM pgrst_args LIMIT 1)")
+        )
+
+    fmtArgs :: (PgArg -> SqlFragment) -> SqlFragment
+    fmtArgs argFrag = intercalate ", " ((\a -> pgFmtIdent (pgaName a) <> argFrag a) <$> pgArgs)
+
+    sourceBody :: SqlFragment
+    sourceBody
+      | paramsAsMulitpleObjects =
+          if returnsScalar
+            then "SELECT " <> callIt <> " AS pgrst_scalar FROM pgrst_args"
+            else unwords [
+              "SELECT pgrst_lat_args.*",
+              "FROM pgrst_args,",
+              "LATERAL ( SELECT * FROM " <> callIt <> " ) pgrst_lat_args" ]
```

Comment:
again nitpick for readability, I hope you do not mind
``` 
else unwords [ "SELECT pgrst_lat_args.*"
                      , "FROM pgrst_args,"
                      , "LATERAL ( SELECT * FROM " <> callIt <> " ) pgrst_lat_args"
                      ]
```

Reply:
I think here you mean aligning to comma first(your snippet is somehow misaligned). I agree, I'll change it.

Reply:
Correct, sorry for the unclear comment. I still cannot get the commas to align, for some reason.